Helios/Interfaces/OpenMacroBoard/MacroBoardInterface.cs

```diff
@@ -0,0 +1,499 @@
+            foreach (MacroBoardButton button in DeckButtons) {
+                writer.WriteStartElement("Button");
+                writer.WriteElementString("Row", button.Row.ToString());
+                writer.WriteElementString("Column", button.Column.ToString());
+                writer.WriteElementString("Text", button.Text);
+                writer.WriteElementString("BackgroundImageUri", button.BackgroundImageUri);
+                writer.WriteElementString("BackgroundImageEnabled", button.BackgroundImageEnabled.ToString());
+                writer.WriteEndElement();
+            }
+            writer.WriteEndElement();
+
+        }
+
+        private void MacroBoardInterface_PropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            if (e.PropertyName == "Model")
+            {
+                Triggers.Remove(_pressedTriggers);
+                Triggers.Remove(_releasedTriggers);
+
+                InitializeTriggers();
+                InitializeButtons();
+            }
+
+            if (e.PropertyName == "Brightness" && IsBoardConnected)
+            {
+                _board?.SetBrightness(Brightness);
+            }
+        }
+
+        private void SendUpdatedButtonImage(int row, int column, RenderTargetBitmap img)
+        {
+            var buttonId = row * ColumnCount + column;
+
+            System.Drawing.Bitmap bitmap;
+
+            //Convert buttonConfig.ButtonImage to a Bitmap
+            using (MemoryStream outStream = new MemoryStream())
+            {
+                BitmapEncoder bEnc = new BmpBitmapEncoder();
+                bEnc.Frames.Add(BitmapFrame.Create(img));
+
+                bEnc.Save(outStream);
+                bitmap = new System.Drawing.Bitmap(outStream);
+                _board.SetKeyBitmap(buttonId, KeyBitmap.Create.FromBitmap(bitmap.Clone(new System.Drawing.Rectangle(0, 0, 72, 72),
+                    System.Drawing.Imaging.PixelFormat.Format32bppArgb)));
+            }
+
+        }
+
+        internal void StreamDeckInterface_ButtonImageChanged(object sender, PropertyChangedEventArgs e)
+        {
+            var buttonConfig = sender as MacroBoardButton;
+
+            if (e.PropertyName == "ButtonImage")
+            {
+                SendUpdatedButtonImage(buttonConfig.Row, buttonConfig.Column, buttonConfig.ButtonImage);
+            }
+
+        }
+
+        internal void Board_KeyStateChanged(object sender, KeyEventArgs e)
+        {
+            MacroBoardButton changedButton = _buttons[e.Key];
+            changedButton.IsPressed = e.IsDown;
+            Console.WriteLine($"Button is pressed: {e.IsDown}");
+
+            HeliosTrigger trigger;
+
+            if (e.IsDown)
+            {
+                trigger = _pressedTriggers[e.Key] as HeliosTrigger;
+            } 
+            else
+            {
+                trigger = _releasedTriggers[e.Key] as HeliosTrigger;
+            }
+
+            if (!Application.Current.Dispatcher.CheckAccess())
```

Comment:
is the KeyStateChanged event on the UI main thread, or is it not?   This should always be either one or the other, right?   I realize this is the canonical MS style, but it would be helpful for the reader to know which case is actually used.

Since your event rate is super low, it is ok that you don't specify the priority level at which you dispatch.  Just note that for other interfaces you might write in the future, selecting the right priority makes a huge difference.

Reply:
Yep, that KeyStateChanged event is done off the main UI thread; the Invoke call prevents an `InvalidOperationException` when you do UI stuff in those event handlers. I can definitely add a comment to reflect that. It might also be worth looking into whether that's something that is worth looking at upstream.

Reply:
Not completely sure if there are cases where it _will_ be on the UI thread, so I put that check there just in case. I might remove it eventually.